widget/textgrid.go

```diff
@@ -341,17 +378,239 @@ func nextTab(column int, tabWidth int) int {
+type textGridContent struct {
+	BaseWidget
+	text *TextGrid
+
+	rows     int
 	cellSize fyne.Size
-	objects  []fyne.CanvasObject
-	current  fyne.Canvas
 }
 
-func (t *textGridRenderer) appendTextCell(str rune) {
+func newTextGridContent(t *TextGrid) *textGridContent {
+	grid := &textGridContent{text: t}
+	grid.ExtendBaseWidget(grid)
+	return grid
+}
+
+// CreateRenderer is a private method to Fyne which links this widget to its renderer
+func (t *textGridContent) CreateRenderer() fyne.WidgetRenderer {
+	r := &textGridContentRenderer{text: t}
+
+	r.updateCellSize()
+	t.text.scroll.OnScrolled = func(_ fyne.Position) {
+		r.addRowsIfRequired()
+		r.Layout(t.Size())
+	}
+	return r
+}
+
+func (t *textGridContent) refreshCell(row, col int) {
+	rows := cache.Renderer(t).Objects()
+	if row >= len(rows)-1 {
+		return
+	}
+	wid := rows[row].(*textGridRow)
+	r := cache.Renderer(wid).(*textGridRowRenderer)
+	r.refreshCell(col)
+}
+
+func (t *textGridContentRenderer) updateGridSize(size fyne.Size) {
+	bufRows := len(t.text.text.Rows)
+	bufCols := 0
+	for _, row := range t.text.text.Rows {
+		bufCols = int(math.Max(float64(bufCols), float64(len(row.Cells))))
+	}
+	sizeRows := math.Floor(float64(size.Height) / float64(t.text.cellSize.Height))
+
+	t.text.rows = int(math.Max(sizeRows, float64(bufRows)))
+	t.addRowsIfRequired()
+}
+
+type textGridContentRenderer struct {
+	text *textGridContent
+
+	visible  []fyne.CanvasObject
+	itemPool async.Pool[*textGridRow]
+}
+
+func (t *textGridContentRenderer) Destroy() {
+}
+
+func (t *textGridContentRenderer) Layout(s fyne.Size) {
+	size := fyne.NewSize(s.Width, t.text.cellSize.Height)
+	t.updateGridSize(s)
+
+	for _, o := range t.visible {
+		o.Move(fyne.NewPos(0, float32(o.(*textGridRow).row)*t.text.cellSize.Height))
+		o.Resize(size)
+	}
+}
+
+func (t *textGridContentRenderer) MinSize() fyne.Size {
+	longestRow := float32(0)
+	for _, row := range t.text.text.Rows {
+		longestRow = fyne.Max(longestRow, float32(len(row.Cells)))
+	}
+	return fyne.NewSize(t.text.cellSize.Width*longestRow,
+		t.text.cellSize.Height*float32(len(t.text.text.Rows)))
+}
+
+func (t *textGridContentRenderer) Objects() []fyne.CanvasObject {
+	return t.visible
+}
+
+func (t *textGridContentRenderer) Refresh() {
+	// theme could change text size
+	t.updateCellSize()
+	t.updateGridSize(t.text.text.Size())
+
+	for _, o := range t.visible {
+		o.Refresh()
+	}
+}
+
+func (t *textGridContentRenderer) addRowsIfRequired() {
+	start := 0
+	end := t.text.rows
+	if t.text.text.Scroll == widget.ScrollBoth || t.text.text.Scroll == widget.ScrollVerticalOnly {
+		off := t.text.text.scroll.Offset.Y
+		start = int(math.Floor(float64(off / t.text.cellSize.Height)))
+
+		off += t.text.text.Size().Height
+		end = int(math.Ceil(float64(off / t.text.cellSize.Height)))
+	}
+
+	var toRemove []*textGridRow
+	for _, row := range t.visible {
+		if row.(*textGridRow).row < start || row.(*textGridRow).row > end {
+			toRemove = append(toRemove, row.(*textGridRow))
+			break
```

Comment:
It looks like `toRemove` is only ever going to be of length one? Seems strange.

Reply:
Good catch